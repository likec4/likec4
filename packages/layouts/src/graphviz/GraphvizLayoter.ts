import { type ComputedView, type DiagramView, isComputedDynamicView } from '@likec4/core'
import logger from '@likec4/log'
import { applyManualLayout } from '../manual/applyManualLayout'
import { DynamicViewPrinter } from './DynamicViewPrinter'
import { ElementViewPrinter } from './ElementViewPrinter'
import { parseGraphvizJson } from './GraphvizParser'
import type { DotSource } from './types'

export interface GraphvizPort {
  unflatten(dot: DotSource): Promise<DotSource>
  acyclic(dot: DotSource): Promise<DotSource>
  layoutJson(dot: DotSource): Promise<string>
  svg(dot: DotSource): Promise<string>
}

const getPrinter = (computedView: ComputedView) =>
  isComputedDynamicView(computedView) ? new DynamicViewPrinter(computedView) : new ElementViewPrinter(computedView)

export type LayoutResult = {
  dot: DotSource
  diagram: DiagramView
}
export class GraphvizLayouter {
  constructor(private graphviz: GraphvizPort) {}

  get port() {
    return this.graphviz
  }

  changePort(graphviz: GraphvizPort) {
    this.graphviz = graphviz
  }

  async layout(view: ComputedView): Promise<LayoutResult> {
    let dot = await this.dot(view)
    const rawjson = await this.graphviz.layoutJson(dot)
    let diagram = parseGraphvizJson(rawjson, view)

    if (view.manualLayout) {
      const result = applyManualLayout(diagram, view.manualLayout)
      if (result.diagram) {
        diagram = result.diagram
      } else {
        // apply manual layout if only new diagram has some nodes
        // from the previous layout
        if (result.relayout.nodes.length > 0) {
          const printer = getPrinter(view)
          // TODO: apply manual layout fails when there are edges with compounds
          if (printer.hasEdgesWithCompounds) {
            // edges with coumpoudns are using _.ltail, _.lhead
            // This is not supported by FDP
            logger.error(`Manual layout for view ${view.id} is ignored, as edges with compounds are not supported`)
          } else {
            printer.applyManualLayout(result.relayout)
            const rawjson = await this.graphviz.layoutJson(printer.print())
            diagram = parseGraphvizJson(rawjson, view)
          }
        }
        diagram.hasLayoutDrift = true
      }
    }

    dot = dot
      .split('\n')
      .filter((l) => !(l.includes('margin') && l.includes('50.1'))) // see DotPrinter.ts#L175
      .join('\n') as DotSource
    return { dot, diagram }
  }

  async svg(view: ComputedView) {
    let dot = await this.dot(view)
    dot = dot
      .split('\n')
      .filter((l) => !(l.includes('margin') && l.includes('50.1'))) // see DotPrinter.ts#L175
      .join('\n') as DotSource
    const svg = await this.graphviz.svg(dot)
    return {
      svg,
      dot
    }
  }

  async dot(computedView: ComputedView): Promise<DotSource> {
    const printer = getPrinter(computedView)
    let dot = printer.print()
    if (isComputedDynamicView(computedView)) {
      return dot
    }
    try {
      return await this.graphviz.unflatten(dot)
    } catch (e) {
      logger.warn(`Error during unflatten: ${computedView.id}`, e)
      return dot
    }
  }
}
