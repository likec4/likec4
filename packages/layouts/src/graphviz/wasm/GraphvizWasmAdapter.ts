import { Graphviz } from '@hpcc-js/wasm/graphviz'
import { delay } from '@likec4/core'
import { logger } from '@likec4/log'
import pLimit from 'p-limit'
import type { GraphvizPort } from '../GraphvizLayoter'
import type { DotSource } from '../types'

const limit = pLimit(1)

export class GraphvizWasmAdapter implements GraphvizPort {
  private static _graphviz: Promise<Graphviz> | null = null

  private graphviz(): Promise<Graphviz> {
    return Promise.resolve().then(() => GraphvizWasmAdapter._graphviz ??= Graphviz.load())
  }

  private async attempt<T>(logMessage: string, fn: () => Promise<T>): Promise<T> {
    return await limit(async () => {
      try {
        return await fn()
      } catch (e) {
        logger.error(e)
        logger.error(`FAILED GraphvizWasmAdapter: ${logMessage}`)
        Graphviz.unload()
        GraphvizWasmAdapter._graphviz = null
      }
      logger.warn('Retrying...')
      await delay(50)
      try {
        return await fn()
      } finally {
        Graphviz.unload()
        GraphvizWasmAdapter._graphviz = null
      }
    })
  }

  async unflatten(dot: DotSource): Promise<DotSource> {
    return await this.attempt(`unflatten\n${dot}`, async () => {
      const graphviz = await this.graphviz()
      const unflattened = graphviz.unflatten(dot, 1, false, 3)
      return unflattened.replaceAll(/\t\[/g, ' [').replaceAll(/\t/g, '    ') as DotSource
    })
  }

  async acyclic(dot: DotSource): Promise<DotSource> {
    return await this.attempt(`acyclic\n${dot}`, async () => {
      const graphviz = await this.graphviz()
      const res = graphviz.acyclic(dot, true)
      return res.acyclic ? (res.outFile as DotSource || dot) : dot
    })
  }

  async layoutJson(dot: DotSource): Promise<string> {
    return await this.attempt(`dot\n${dot}`, async () => {
      const graphviz = await this.graphviz()
      return graphviz.layout(dot, 'json', undefined, {
        yInvert: true
      })
    })
  }

  async svg(dot: DotSource): Promise<string> {
    return await this.attempt(`svg\n${dot}`, async () => {
      const graphviz = await this.graphviz()
      return graphviz.layout(dot, 'svg')
    })
  }
}
