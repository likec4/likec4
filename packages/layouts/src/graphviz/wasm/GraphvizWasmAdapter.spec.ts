import type { ComputedView } from '@likec4/core'
import { describe, it } from 'vitest'
import {
  computedAmazonView,
  computedCloud3levels,
  computedCloudView,
  computedIndexView,
  issue577_fail,
  issue577_valid,
  parsedModel,
} from '../__fixtures__'
import { GraphvizLayouter } from '../GraphvizLayoter'
import { GraphvizWasmAdapter } from './GraphvizWasmAdapter'

async function dotLayout(computedView: ComputedView) {
  const graphviz = new GraphvizLayouter(new GraphvizWasmAdapter())
  return (await graphviz.layout({
    specification: parsedModel.specification,
    view: computedView,
  })).diagram
}

describe('GraphvizWasmAdapter:', () => {
  // it('computedIndexView', async ({ expect }) => {
  //   const graphviz = new GraphvizLayouter(new GraphvizWasmAdapter())

  //   const {diagram, dot} = await graphviz.layout(computedCloud3levels)
  //   // expect(diagram).toMatchSnapshot()
  //   expect(dot).toMatchSnapshot()

  //   const nextdot = await graphviz.dot(computedCloud3levels, diagram)

  //   expect(nextdot).toMatchSnapshot()
  // })

  it('computedIndexView', async ({ expect }) => {
    const diagram = await dotLayout(computedIndexView)
    expect(diagram).toMatchSnapshot()
  })

  it('computedAmazonView', async ({ expect }) => {
    const diagram = await dotLayout(computedAmazonView)
    expect(diagram).toMatchSnapshot()
  })

  it('computedCloud3levels', async ({ expect }) => {
    const diagram = await dotLayout(computedCloud3levels)
    expect(diagram).toMatchSnapshot()
  })

  it('computedCloudView', async ({ expect }) => {
    const diagram = await dotLayout(computedCloudView)
    expect(diagram).toMatchSnapshot()
  })

  it('reproduce #577', async ({ expect }) => {
    // was failing with invalid URL
    const diagram = await dotLayout(issue577_fail)
    expect(diagram).toBeDefined()
    expect(diagram.nodes[0]?.icon).toEqual('https://icons/aws%20&%20CloudFront.svg')

    // was valid
    expect(await dotLayout(issue577_valid)).toBeDefined()
  })
})
