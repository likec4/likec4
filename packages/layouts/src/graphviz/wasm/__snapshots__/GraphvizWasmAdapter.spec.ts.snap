// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`GraphvizWasmAdapter: > computedAmazonView 1`] = `
{
  "__": "element",
  "autoLayout": "TB",
  "description": null,
  "edges": [
    {
      "dotpos": "s,359.91,151.2 e,567.29,151.2 370.78,151.2 430.46,151.2 497.41,151.2 557.02,151.2",
      "id": "cloud.backend:amazon.s3",
      "label": "persists artifacts",
      "labelBBox": {
        "height": 18,
        "width": 106,
        "x": 411,
        "y": 125,
      },
      "parent": null,
      "points": [
        [
          371,
          152,
        ],
        [
          431,
          152,
        ],
        [
          498,
          152,
        ],
        [
          558,
          152,
        ],
      ],
      "relations": [
        "cloud.backend.storage:amazon.s3",
      ],
      "source": "cloud.backend",
      "tail": "odiamond",
      "target": "amazon.s3",
    },
  ],
  "hash": "b0893852a41efeb3d6e941dbe5221bde3d9a048a",
  "height": 282,
  "id": "amazon",
  "links": null,
  "nodes": [
    {
      "children": [
        "cloud.backend",
      ],
      "color": "primary",
      "depth": 1,
      "description": null,
      "height": 266,
      "id": "cloud",
      "inEdges": [],
      "kind": "system",
      "labelBBox": {
        "height": 15,
        "width": 44,
        "x": 6,
        "y": 0,
      },
      "level": 0,
      "links": null,
      "outEdges": [
        "cloud.backend:amazon.s3",
      ],
      "parent": null,
      "position": [
        8,
        8,
      ],
      "shape": "rectangle",
      "style": {},
      "tags": null,
      "technology": null,
      "title": "cloud",
      "width": 384,
    },
    {
      "children": [],
      "color": "primary",
      "description": null,
      "height": 180,
      "id": "cloud.backend",
      "inEdges": [],
      "kind": "component",
      "labelBBox": {
        "height": 23,
        "width": 78,
        "x": 123,
        "y": 74,
      },
      "level": 1,
      "links": null,
      "outEdges": [
        "cloud.backend:amazon.s3",
      ],
      "parent": "cloud",
      "position": [
        39,
        62,
      ],
      "shape": "rectangle",
      "style": {},
      "tags": null,
      "technology": null,
      "title": "Backend",
      "width": 321,
    },
    {
      "children": [
        "amazon.s3",
      ],
      "color": "primary",
      "depth": 1,
      "description": "Amazon is a cloud provider",
      "height": 266,
      "id": "amazon",
      "inEdges": [
        "cloud.backend:amazon.s3",
      ],
      "kind": "system",
      "labelBBox": {
        "height": 15,
        "width": 53,
        "x": 6,
        "y": 0,
      },
      "level": 0,
      "links": null,
      "outEdges": [],
      "parent": null,
      "position": [
        535,
        8,
      ],
      "shape": "rectangle",
      "style": {},
      "tags": null,
      "technology": null,
      "title": "Amazon",
      "width": 384,
    },
    {
      "children": [],
      "color": "primary",
      "description": "S3 is a storage service",
      "height": 180,
      "id": "amazon.s3",
      "inEdges": [
        "cloud.backend:amazon.s3",
      ],
      "kind": "component",
      "labelBBox": {
        "height": 47,
        "width": 146,
        "x": 89,
        "y": 63,
      },
      "level": 1,
      "links": null,
      "outEdges": [],
      "parent": "amazon",
      "position": [
        566,
        62,
      ],
      "shape": "rectangle",
      "style": {},
      "tags": null,
      "technology": null,
      "title": "S3",
      "width": 321,
    },
  ],
  "tags": null,
  "title": "",
  "viewOf": "amazon",
  "width": 927,
}
`;

exports[`GraphvizWasmAdapter: > computedCloud3levels 1`] = `
{
  "__": "element",
  "autoLayout": "TB",
  "description": null,
  "edges": [
    {
      "dotpos": "e,216,319.02 216,179.72 216,219.83 216,267.38 216,308.67",
      "id": "support:cloud.frontend.adminPanel",
      "label": "manages",
      "labelBBox": {
        "height": 18,
        "width": 62,
        "x": 217,
        "y": 235,
      },
      "parent": null,
      "points": [
        [
          216,
          180,
        ],
        [
          216,
          220,
        ],
        [
          216,
          268,
        ],
        [
          216,
          309,
        ],
      ],
      "relations": [
        "support:cloud.frontend.adminPanel",
      ],
      "source": "support",
      "target": "cloud.frontend.adminPanel",
    },
    {
      "dotpos": "e,639,319.02 639,179.72 639,219.83 639,267.38 639,308.67",
      "id": "customer:cloud.frontend.dashboard",
      "label": "opens in browser",
      "labelBBox": {
        "height": 18,
        "width": 111,
        "x": 640,
        "y": 235,
      },
      "parent": null,
      "points": [
        [
          639,
          180,
        ],
        [
          639,
          220,
        ],
        [
          639,
          268,
        ],
        [
          639,
          309,
        ],
      ],
      "relations": [
        "customer:cloud.frontend.dashboard",
      ],
      "source": "customer",
      "target": "cloud.frontend.dashboard",
    },
    {
      "dotpos": "e,227.73,654.81 220.25,498.58 222.4,543.47 225.02,598.13 227.24,644.45",
      "id": "cloud.frontend.adminPanel:cloud.backend.graphql",
      "label": "fetches data in zero trust network
with sso authentification",
      "labelBBox": {
        "height": 35,
        "width": 211,
        "x": 225,
        "y": 554,
      },
      "parent": "cloud",
      "points": [
        [
          221,
          499,
        ],
        [
          223,
          544,
        ],
        [
          226,
          599,
        ],
        [
          228,
          645,
        ],
      ],
      "relations": [
        "cloud.frontend.adminPanel:cloud.backend.graphql",
      ],
      "source": "cloud.frontend.adminPanel",
      "target": "cloud.backend.graphql",
    },
    {
      "dotpos": "e,354.67,654.86 566.43,498.82 536.34,531.68 499.48,567.29 461,593.6 453.44,598.77 450.05,597.23 442,601.6 415.83,615.79 388.84,632.4 363.28,649.17",
      "id": "cloud.frontend.dashboard:cloud.backend.graphql",
      "label": "fetches data",
      "labelBBox": {
        "height": 18,
        "width": 81,
        "x": 511,
        "y": 562,
      },
      "parent": "cloud",
      "points": [
        [
          567,
          499,
        ],
        [
          537,
          532,
        ],
        [
          500,
          568,
        ],
        [
          461,
          594,
        ],
        [
          454,
          599,
        ],
        [
          451,
          598,
        ],
        [
          442,
          602,
        ],
        [
          416,
          616,
        ],
        [
          389,
          633,
        ],
        [
          364,
          650,
        ],
      ],
      "relations": [
        "cloud.frontend.dashboard:cloud.backend.graphql",
      ],
      "source": "cloud.frontend.dashboard",
      "target": "cloud.backend.graphql",
    },
    {
      "dotpos": "e,491.36,744.8 391.93,744.8 421.58,744.8 451.23,744.8 480.87,744.8",
      "id": "cloud.backend.graphql:cloud.backend.storage",
      "label": null,
      "labelBBox": null,
      "parent": "cloud.backend",
      "points": [
        [
          392,
          745,
        ],
        [
          422,
          745,
        ],
        [
          452,
          745,
        ],
        [
          481,
          745,
        ],
      ],
      "relations": [
        "cloud.backend.graphql:cloud.backend.storage",
      ],
      "source": "cloud.backend.graphql",
      "target": "cloud.backend.storage",
    },
    {
      "dotpos": "s,814.7,744.8 e,1020.4,744.8 825.51,744.8 884.82,744.8 951.15,744.8 1010.2,744.8",
      "id": "cloud.backend.storage:amazon",
      "label": "persists artifacts",
      "labelBBox": {
        "height": 18,
        "width": 106,
        "x": 865,
        "y": 719,
      },
      "parent": null,
      "points": [
        [
          826,
          745,
        ],
        [
          885,
          745,
        ],
        [
          952,
          745,
        ],
        [
          1011,
          745,
        ],
      ],
      "relations": [
        "cloud.backend.storage:amazon.s3",
      ],
      "source": "cloud.backend.storage",
      "tail": "odiamond",
      "target": "amazon",
    },
  ],
  "hash": "d251757de7f97a6d12026d6cb711903508a5e227",
  "height": 907,
  "id": "cloud3levels",
  "links": null,
  "nodes": [
    {
      "children": [],
      "color": "primary",
      "description": "Support engineers are responsible for supporting customers",
      "height": 180,
      "id": "support",
      "inEdges": [],
      "kind": "actor",
      "labelBBox": {
        "height": 64,
        "width": 239,
        "x": 42,
        "y": 55,
      },
      "level": 0,
      "links": null,
      "outEdges": [
        "support:cloud.frontend.adminPanel",
      ],
      "parent": null,
      "position": [
        55,
        0,
      ],
      "shape": "person",
      "style": {},
      "tags": null,
      "technology": null,
      "title": "Support Engineer",
      "width": 321,
    },
    {
      "children": [],
      "color": "primary",
      "description": null,
      "height": 180,
      "id": "customer",
      "inEdges": [],
      "kind": "actor",
      "labelBBox": {
        "height": 23,
        "width": 83,
        "x": 120,
        "y": 75,
      },
      "level": 0,
      "links": null,
      "outEdges": [
        "customer:cloud.frontend.dashboard",
      ],
      "parent": null,
      "position": [
        478,
        0,
      ],
      "shape": "person",
      "style": {},
      "tags": null,
      "technology": null,
      "title": "customer",
      "width": 321,
    },
    {
      "children": [
        "cloud.frontend.adminPanel",
        "cloud.frontend.dashboard",
        "cloud.backend",
      ],
      "color": "primary",
      "depth": 2,
      "description": null,
      "height": 633,
      "id": "cloud",
      "inEdges": [
        "customer:cloud.frontend.dashboard",
        "support:cloud.frontend.adminPanel",
      ],
      "kind": "system",
      "labelBBox": {
        "height": 15,
        "width": 44,
        "x": 6,
        "y": 0,
      },
      "level": 0,
      "links": null,
      "outEdges": [
        "cloud.backend.storage:amazon",
      ],
      "parent": null,
      "position": [
        8,
        266,
      ],
      "shape": "rectangle",
      "style": {},
      "tags": null,
      "technology": null,
      "title": "cloud",
      "width": 870,
    },
    {
      "children": [],
      "color": "primary",
      "description": "The admin panel is a webapp that allows support staff to manage customer data.",
      "height": 180,
      "id": "cloud.frontend.adminPanel",
      "inEdges": [
        "support:cloud.frontend.adminPanel",
      ],
      "kind": "component",
      "labelBBox": {
        "height": 64,
        "width": 257,
        "x": 33,
        "y": 55,
      },
      "level": 1,
      "links": null,
      "outEdges": [
        "cloud.frontend.adminPanel:cloud.backend.graphql",
      ],
      "parent": "cloud",
      "position": [
        55,
        319,
      ],
      "shape": "rectangle",
      "style": {},
      "tags": null,
      "technology": null,
      "title": "Admin Panel Webapp",
      "width": 321,
    },
    {
      "children": [],
      "color": "primary",
      "description": "The customer dashboard is a webapp that allows customers to view their data.",
      "height": 180,
      "id": "cloud.frontend.dashboard",
      "inEdges": [
        "customer:cloud.frontend.dashboard",
      ],
      "kind": "component",
      "labelBBox": {
        "height": 64,
        "width": 266,
        "x": 31,
        "y": 55,
      },
      "level": 1,
      "links": null,
      "outEdges": [
        "cloud.frontend.dashboard:cloud.backend.graphql",
      ],
      "parent": "cloud",
      "position": [
        476,
        319,
      ],
      "shape": "rectangle",
      "style": {},
      "tags": null,
      "technology": null,
      "title": "Customer Dashboard Webapp",
      "width": 326,
    },
    {
      "children": [
        "cloud.backend.graphql",
        "cloud.backend.storage",
      ],
      "color": "primary",
      "depth": 1,
      "description": null,
      "height": 265,
      "id": "cloud.backend",
      "inEdges": [
        "cloud.frontend.dashboard:cloud.backend.graphql",
        "cloud.frontend.adminPanel:cloud.backend.graphql",
      ],
      "kind": "component",
      "labelBBox": {
        "height": 15,
        "width": 59,
        "x": 6,
        "y": 0,
      },
      "level": 1,
      "links": null,
      "outEdges": [
        "cloud.backend.storage:amazon",
      ],
      "parent": "cloud",
      "position": [
        40,
        602,
      ],
      "shape": "rectangle",
      "style": {},
      "tags": null,
      "technology": null,
      "title": "Backend",
      "width": 806,
    },
    {
      "children": [],
      "color": "primary",
      "description": "Component that allows to query data via GraphQL.",
      "height": 180,
      "id": "cloud.backend.graphql",
      "inEdges": [
        "cloud.frontend.dashboard:cloud.backend.graphql",
        "cloud.frontend.adminPanel:cloud.backend.graphql",
      ],
      "kind": "component",
      "labelBBox": {
        "height": 64,
        "width": 254,
        "x": 35,
        "y": 55,
      },
      "level": 2,
      "links": null,
      "outEdges": [
        "cloud.backend.graphql:cloud.backend.storage",
      ],
      "parent": "cloud.backend",
      "position": [
        71,
        655,
      ],
      "shape": "rectangle",
      "style": {},
      "tags": null,
      "technology": null,
      "title": "Graphql API",
      "width": 321,
    },
    {
      "children": [],
      "color": "primary",
      "description": "The backend storage is a component that stores data.",
      "height": 180,
      "id": "cloud.backend.storage",
      "inEdges": [
        "cloud.backend.graphql:cloud.backend.storage",
      ],
      "kind": "component",
      "labelBBox": {
        "height": 64,
        "width": 262,
        "x": 31,
        "y": 55,
      },
      "level": 2,
      "links": null,
      "outEdges": [
        "cloud.backend.storage:amazon",
      ],
      "parent": "cloud.backend",
      "position": [
        492,
        655,
      ],
      "shape": "storage",
      "style": {},
      "tags": null,
      "technology": null,
      "title": "Backend Storage",
      "width": 322,
    },
    {
      "children": [],
      "color": "primary",
      "description": "Amazon is a cloud provider",
      "height": 180,
      "id": "amazon",
      "inEdges": [
        "cloud.backend.storage:amazon",
      ],
      "kind": "system",
      "labelBBox": {
        "height": 47,
        "width": 173,
        "x": 75,
        "y": 64,
      },
      "level": 0,
      "links": null,
      "outEdges": [],
      "parent": null,
      "position": [
        1019,
        655,
      ],
      "shape": "rectangle",
      "style": {},
      "tags": null,
      "technology": null,
      "title": "Amazon",
      "width": 321,
    },
  ],
  "tags": null,
  "title": "",
  "viewOf": "cloud",
  "width": 1340,
}
`;

exports[`GraphvizWasmAdapter: > computedCloudView 1`] = `
{
  "__": "element",
  "autoLayout": "TB",
  "description": null,
  "edges": [
    {
      "dotpos": "e,311.08,319.02 218.8,179.72 245.76,220.42 277.8,268.79 305.43,310.5",
      "id": "customer:cloud.frontend",
      "label": "opens in browser",
      "labelBBox": {
        "height": 18,
        "width": 111,
        "x": 267,
        "y": 235,
      },
      "parent": null,
      "points": [
        [
          219,
          180,
        ],
        [
          246,
          221,
        ],
        [
          278,
          269,
        ],
        [
          306,
          311,
        ],
      ],
      "relations": [
        "customer:cloud.frontend.dashboard",
      ],
      "source": "customer",
      "target": "cloud.frontend",
    },
    {
      "dotpos": "e,428.96,319.02 521.24,179.72 494.28,220.42 462.24,268.79 434.61,310.5",
      "id": "support:cloud.frontend",
      "label": "manages",
      "labelBBox": {
        "height": 18,
        "width": 62,
        "x": 483,
        "y": 235,
      },
      "parent": null,
      "points": [
        [
          522,
          180,
        ],
        [
          495,
          221,
        ],
        [
          463,
          269,
        ],
        [
          435,
          311,
        ],
      ],
      "relations": [
        "support:cloud.frontend.adminPanel",
      ],
      "source": "support",
      "target": "cloud.frontend",
    },
    {
      "dotpos": "e,657.06,409 529.82,409 567.78,409 608.41,409 646.82,409",
      "id": "cloud.frontend:cloud.backend",
      "label": "[...]",
      "labelBBox": {
        "height": 18,
        "width": 25,
        "x": 582,
        "y": 382,
      },
      "parent": "cloud",
      "points": [
        [
          530,
          409,
        ],
        [
          568,
          409,
        ],
        [
          609,
          409,
        ],
        [
          647,
          409,
        ],
      ],
      "relations": [
        "cloud.frontend.dashboard:cloud.backend.graphql",
        "cloud.frontend.adminPanel:cloud.backend.graphql",
      ],
      "source": "cloud.frontend",
      "target": "cloud.backend",
    },
    {
      "dotpos": "s,976.93,409 e,1184.3,409 987.8,409 1047.5,409 1114.4,409 1174,409",
      "id": "cloud.backend:amazon",
      "label": "persists artifacts",
      "labelBBox": {
        "height": 18,
        "width": 106,
        "x": 1028,
        "y": 383,
      },
      "parent": null,
      "points": [
        [
          988,
          409,
        ],
        [
          1048,
          409,
        ],
        [
          1115,
          409,
        ],
        [
          1175,
          409,
        ],
      ],
      "relations": [
        "cloud.backend.storage:amazon.s3",
      ],
      "source": "cloud.backend",
      "tail": "odiamond",
      "target": "amazon",
    },
  ],
  "hash": "a2aecd4a69928a2ba4e12c37df5eb263ef503e31",
  "height": 539,
  "id": "cloudView",
  "links": null,
  "nodes": [
    {
      "children": [],
      "color": "primary",
      "description": null,
      "height": 180,
      "id": "customer",
      "inEdges": [],
      "kind": "actor",
      "labelBBox": {
        "height": 23,
        "width": 83,
        "x": 119,
        "y": 75,
      },
      "level": 0,
      "links": null,
      "outEdges": [
        "customer:cloud.frontend",
      ],
      "parent": null,
      "position": [
        0,
        0,
      ],
      "shape": "person",
      "style": {},
      "tags": null,
      "technology": null,
      "title": "customer",
      "width": 321,
    },
    {
      "children": [],
      "color": "primary",
      "description": "Support engineers are responsible for supporting customers",
      "height": 180,
      "id": "support",
      "inEdges": [],
      "kind": "actor",
      "labelBBox": {
        "height": 64,
        "width": 239,
        "x": 41,
        "y": 55,
      },
      "level": 0,
      "links": null,
      "outEdges": [
        "support:cloud.frontend",
      ],
      "parent": null,
      "position": [
        420,
        0,
      ],
      "shape": "person",
      "style": {},
      "tags": null,
      "technology": null,
      "title": "Support Engineer",
      "width": 321,
    },
    {
      "children": [
        "cloud.frontend",
        "cloud.backend",
      ],
      "color": "primary",
      "depth": 1,
      "description": null,
      "height": 265,
      "id": "cloud",
      "inEdges": [
        "support:cloud.frontend",
        "customer:cloud.frontend",
      ],
      "kind": "system",
      "labelBBox": {
        "height": 15,
        "width": 44,
        "x": 6,
        "y": 0,
      },
      "level": 0,
      "links": null,
      "outEdges": [
        "cloud.backend:amazon",
      ],
      "parent": null,
      "position": [
        179,
        266,
      ],
      "shape": "rectangle",
      "style": {},
      "tags": null,
      "technology": null,
      "title": "cloud",
      "width": 830,
    },
    {
      "children": [],
      "color": "primary",
      "description": null,
      "height": 180,
      "id": "cloud.frontend",
      "inEdges": [
        "support:cloud.frontend",
        "customer:cloud.frontend",
      ],
      "kind": "component",
      "labelBBox": {
        "height": 23,
        "width": 81,
        "x": 120,
        "y": 75,
      },
      "level": 1,
      "links": null,
      "outEdges": [
        "cloud.frontend:cloud.backend",
      ],
      "parent": "cloud",
      "position": [
        210,
        319,
      ],
      "shape": "browser",
      "style": {},
      "tags": null,
      "technology": null,
      "title": "Frontend",
      "width": 321,
    },
    {
      "children": [],
      "color": "primary",
      "description": null,
      "height": 180,
      "id": "cloud.backend",
      "inEdges": [
        "cloud.frontend:cloud.backend",
      ],
      "kind": "component",
      "labelBBox": {
        "height": 23,
        "width": 78,
        "x": 122,
        "y": 75,
      },
      "level": 1,
      "links": null,
      "outEdges": [
        "cloud.backend:amazon",
      ],
      "parent": "cloud",
      "position": [
        657,
        319,
      ],
      "shape": "rectangle",
      "style": {},
      "tags": null,
      "technology": null,
      "title": "Backend",
      "width": 321,
    },
    {
      "children": [],
      "color": "primary",
      "description": "Amazon is a cloud provider",
      "height": 180,
      "id": "amazon",
      "inEdges": [
        "cloud.backend:amazon",
      ],
      "kind": "system",
      "labelBBox": {
        "height": 47,
        "width": 173,
        "x": 75,
        "y": 64,
      },
      "level": 0,
      "links": null,
      "outEdges": [],
      "parent": null,
      "position": [
        1183,
        319,
      ],
      "shape": "rectangle",
      "style": {},
      "tags": null,
      "technology": null,
      "title": "Amazon",
      "width": 321,
    },
  ],
  "tags": null,
  "title": "",
  "viewOf": "cloud",
  "width": 1504,
}
`;

exports[`GraphvizWasmAdapter: > computedIndexView 1`] = `
{
  "__": "element",
  "autoLayout": "TB",
  "description": null,
  "edges": [
    {
      "dotpos": "e,310.04,313.04 219.92,179.65 246.23,218.59 277.22,264.45 304.18,304.35",
      "id": "customer:cloud",
      "label": "opens in browser",
      "labelBBox": {
        "height": 18,
        "width": 111,
        "x": 274,
        "y": 235,
      },
      "parent": null,
      "points": [
        [
          220,
          180,
        ],
        [
          247,
          219,
        ],
        [
          278,
          265,
        ],
        [
          305,
          305,
        ],
      ],
      "relations": [
        "customer:cloud.frontend.dashboard",
      ],
      "source": "customer",
      "target": "cloud",
    },
    {
      "dotpos": "e,430,313.04 520.12,179.65 493.81,218.59 462.82,264.45 435.86,304.35",
      "id": "support:cloud",
      "label": "manages",
      "labelBBox": {
        "height": 18,
        "width": 62,
        "x": 483,
        "y": 235,
      },
      "parent": null,
      "points": [
        [
          521,
          180,
        ],
        [
          494,
          219,
        ],
        [
          463,
          265,
        ],
        [
          436,
          305,
        ],
      ],
      "relations": [
        "support:cloud.frontend.adminPanel",
      ],
      "source": "support",
      "target": "cloud",
    },
    {
      "dotpos": "s,529.93,402.8 e,737.31,402.8 540.8,402.8 600.48,402.8 667.43,402.8 727.04,402.8",
      "id": "cloud:amazon",
      "label": "persists artifacts",
      "labelBBox": {
        "height": 18,
        "width": 106,
        "x": 581,
        "y": 377,
      },
      "parent": null,
      "points": [
        [
          541,
          403,
        ],
        [
          601,
          403,
        ],
        [
          668,
          403,
        ],
        [
          728,
          403,
        ],
      ],
      "relations": [
        "cloud.backend.storage:amazon.s3",
      ],
      "source": "cloud",
      "tail": "odiamond",
      "target": "amazon",
    },
  ],
  "hash": "1cb03738dd683e790ee2267d046851d00b0d9c1a",
  "height": 493,
  "id": "index",
  "links": null,
  "nodes": [
    {
      "children": [],
      "color": "primary",
      "description": null,
      "height": 180,
      "id": "customer",
      "inEdges": [],
      "kind": "actor",
      "labelBBox": {
        "height": 23,
        "width": 83,
        "x": 119,
        "y": 75,
      },
      "level": 0,
      "links": null,
      "outEdges": [
        "customer:cloud",
      ],
      "parent": null,
      "position": [
        0,
        0,
      ],
      "shape": "person",
      "style": {},
      "tags": null,
      "technology": null,
      "title": "customer",
      "width": 321,
    },
    {
      "children": [],
      "color": "primary",
      "description": "Support engineers are responsible for supporting customers",
      "height": 180,
      "id": "support",
      "inEdges": [],
      "kind": "actor",
      "labelBBox": {
        "height": 64,
        "width": 239,
        "x": 41,
        "y": 55,
      },
      "level": 0,
      "links": null,
      "outEdges": [
        "support:cloud",
      ],
      "parent": null,
      "position": [
        420,
        0,
      ],
      "shape": "person",
      "style": {},
      "tags": null,
      "technology": null,
      "title": "Support Engineer",
      "width": 321,
    },
    {
      "children": [],
      "color": "primary",
      "description": null,
      "height": 180,
      "id": "cloud",
      "inEdges": [
        "support:cloud",
        "customer:cloud",
      ],
      "kind": "system",
      "labelBBox": {
        "height": 23,
        "width": 50,
        "x": 136,
        "y": 75,
      },
      "level": 0,
      "links": null,
      "outEdges": [
        "cloud:amazon",
      ],
      "parent": null,
      "position": [
        210,
        313,
      ],
      "shape": "rectangle",
      "style": {},
      "tags": null,
      "technology": null,
      "title": "cloud",
      "width": 321,
    },
    {
      "children": [],
      "color": "primary",
      "description": "Amazon is a cloud provider",
      "height": 180,
      "id": "amazon",
      "inEdges": [
        "cloud:amazon",
      ],
      "kind": "system",
      "labelBBox": {
        "height": 47,
        "width": 173,
        "x": 74,
        "y": 64,
      },
      "level": 0,
      "links": null,
      "outEdges": [],
      "parent": null,
      "position": [
        737,
        313,
      ],
      "shape": "rectangle",
      "style": {},
      "tags": null,
      "technology": null,
      "title": "Amazon",
      "width": 321,
    },
  ],
  "tags": null,
  "title": "",
  "width": 1057,
}
`;
